src/qt/treectrl.cpp

```diff
@@ -8,11 +8,438 @@
+                this, &wxQTreeWidget::OnItemClicked);
+        connect(this, &QTreeWidget::itemCollapsed,
+                this, &wxQTreeWidget::OnItemCollapsed);
+        connect(this, &QTreeWidget::itemExpanded,
+                this, &wxQTreeWidget::OnItemExpanded);
+        connect(this, &QTreeWidget::itemChanged,
+                this, &wxQTreeWidget::OnItemChanged);
+        connect(this, &QTreeWidget::iconSizeChanged,
+                this, &wxQTreeWidget::OnIconSizeChanged);
+
+        m_editorFactory.AttachTo(this);
+        setDragEnabled(true);
+        viewport()->setAcceptDrops(true);
+        setDropIndicatorShown(true);
+    }
+
+    wxTextCtrl *GetEditControl()
+    {
+        return m_editorFactory.GetEditControl();
+    }
+
+    void SetItemImage(QTreeWidgetItem *item, int image, wxTreeItemIcon which)
+    {
+        m_imageStates[item][which] = image;
+    }
+
+    int GetItemImage(QTreeWidgetItem *item, wxTreeItemIcon which)
+    {
+        if ( m_imageStates.find(item) == m_imageStates.end() )
+            return 0;
+
+        return m_imageStates[item][which];
+    }
+
+    void SetItemState(QTreeWidgetItem *item, int state)
+    {
+        m_imageStates[item].SetState(state);
+    }
+        
+    int GetItemState(QTreeWidgetItem *item) const
+    {
+        const ImageStateMap::const_iterator i = m_imageStates.find(item);
+        if ( i == m_imageStates.end() )
+            return wxTREE_ITEMSTATE_NONE;
+
+        return i->second.GetState();
+    }
+
+    QPixmap GetPlaceHolderImage() const
+    {
+        return m_placeHolderImage;
+    }
+
+protected:
+    virtual void drawBranches(
+        QPainter *painter,
+        const QRect &rect,
+        const QModelIndex &index
+    ) const wxOVERRIDE
+    {
+        QTreeWidgetItem *item = itemFromIndex(index);
+
+        QTreeWidget::drawBranches(painter, rect, index);
+
+        const int imageIndex = ChooseBestImage(item);
+        if ( imageIndex != -1 )
+        {
+            const wxImageList *imageList = GetHandler()->GetImageList();
+            const wxBitmap bitmap = imageList->GetBitmap(imageIndex);
+            painter->drawPixmap(rect.topRight(), *bitmap.GetHandle());
+        }
+    }
+
+private:
+    void OnCurrentItemChanged(
+        QTreeWidgetItem *current,
+        QTreeWidgetItem *previous
+    )
+    {
+        wxTreeCtrl* treeCtrl = GetHandler();
+
+        wxTreeEvent changingEvent(
+            wxEVT_TREE_SEL_CHANGING,
+            treeCtrl,
+            wxQtConvertTreeItem(current)
+        );
+
+        changingEvent.SetOldItem(wxQtConvertTreeItem(previous));
+        EmitEvent(changingEvent);
+
+        if ( !changingEvent.IsAllowed() )
+        {
+            blockSignals(true);
+            setCurrentItem(previous);
+            blockSignals(false);
+            return;
+        }
+
+        // QT doesn't update the selection until this signal has been
+        // processed. Deferring this event ensures that
```

Comment:
Thanks for the explanation, this is very helpful.

Just out of curiosity, is there no possibility to define a "post-signal" handler in Qt, like GTK+ "after" signals, i.e. something that happens after the default handler executes?